---
title: "Single Cell Figures"
author: "Anneke Kakebeen"
date: "13 Aug 2019"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
editor_options: 
  chunk_output_type: console
---

## Set up for scRNA-Seq Figures
```{r scrnaseq figure set up, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE}
# load packages
library(Seurat)
library(LaCroixColoR)
library(ggplot2)
library(cowplot)
library(tidyr)
library(DT)
library(dplyr)

# load functions
source("~/Desktop/pax6 paper/Final Markdowns/common_source_functions.R")

# Set colors
my.color.neural <-unique( c(lacroix_palette("Berry"), lacroix_palette("CranRaspberry"), lacroix_palette("Tangerine"),        lacroix_palette("PinaFraise")))
condition.color <- c( "#6F0909", "#C0C0C0")
cc_colors <- c((lacroix_palette("Lime"))[c(1,3)], "black")

# Set working directory
setwd("~/Desktop/pax6 paper/Final Markdowns/SingleCell/analysis/")

```



## Read in scRNASeq Objects
```{r load objects, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE}

# Read in pax.all seurat object
pax.all <- readRDS("~/Desktop/pax6 paper/Final Markdowns/SingleCell/UMAP/pax.combined_FINAL.RDS")

# read in pax.neural
pax.neural <- readRDS("~/Desktop/pax6 paper/Final Markdowns/SingleCell/UMAP/pax.neural_FINAL.RDS")

# read in plotting data
pax.all.plotting<- read.table("~/Desktop/pax6 paper/Final Markdowns/SingleCell/UMAP/pax.combined_plotting.txt", sep = "\t")

# factor plotting table for plotting
pax.all.plotting$condition <- factor(pax.all.plotting$condition, levels = c("uninj", "dmso", "iwr"))
pax.all.plotting$Phase <- factor(pax.all.plotting$Phase, levels = c("S", "G2M", "G1"))
```


## Supp 4A: UMAP of all cells colored by individual cell types
```{r supp 4a, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE}

# change identity to cell clsuter
Idents(pax.all) <- pax.all$AK_Names

# plot cells in umap
DimPlot(pax.all, cols = my.color.neural) +
  NoAxes() +
  NoLegend()

```

## Supp 4B: featureplot for neural cell markers
```{r supp 4c, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE}

# set assay
DefaultAssay(pax.all) <- "RNA"
# plot neural markers

p1 <- FeaturePlot(pax.all, features = c("sox2"),  order = TRUE, cols = c("dodgerblue3", "yellow", "forestgreen"))+ NoAxes()+NoLegend()
p2 <- FeaturePlot(pax.all, features = c("neurog1"),  order = TRUE, cols = c("dodgerblue3", "yellow", "forestgreen"))+ NoAxes()+NoLegend()
p3 <- FeaturePlot(pax.all, features = c("neurod1"),  order = TRUE, cols = c("dodgerblue3", "yellow", "forestgreen"))+ NoAxes()+NoLegend()
p4 <- FeaturePlot(pax.all, features = c("elavl4"),  order = TRUE, cols = c("dodgerblue3", "yellow", "forestgreen"))+ NoAxes()+NoLegend()

plot_grid(p1, p2, p3, p4, nrow = 2)

```

## Supp 4C: UMAP of all cells colored by cell classes
```{r supp 4b, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE}

Idents(pax.all) <- pax.all$CellClass
DimPlot(pax.all, label = FALSE, cols = my.color.neural[c(1, 3, 5, 7, 9, 11)]) +
  NoAxes() +
  NoLegend()


```



## Supp 4D: Average expression bar plots of pbx3 and meis1 in all cells
### Bar plots
```{r grn, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE}

# Predicted GRN
" 6hpa_Gene:meis1 -> 24hpa_TF:meis1 -> 72hpa_TF:runx|etv1|klf9"
" 6hpa_Gene:etv1 -> 24hpa_TF:etv1 -> 72hpa_TF:pbx3"

# Define ins and outs for extraction function
Idents(pax.all) <- pax.all$CellClass
average_clusters <- t(as.data.frame(AverageExpression(pax.all, assays = "RNA", use.counts = TRUE, features = c("meis1", "pbx3"))))
outs <- c("meis1_out", "pbx3_out")
genes <- c("meis1", "pbx3")

# Extract average expression data for plotting
seuratTOplot(average_clusters, gene = genes, outs = outs) # run to plot function (found in sourced functions)
objects <- ls(pattern=".*_out") # get all object names made in previous function
objects.list <- lapply(objects, get) # get outs
plot <- do.call(bind_rows, objects.list) # rowbind both gene data sets
# Make metdata columns
plot$treatment <- gsub(plot$condition, pattern = "RNA.", replacement = "") # split condition column

# Factor data to order
plot$gene <- factor(plot$gene, levels = c("pbx3", "meis1")) # order gene row for plot

# # Factor for order of cell types
# plot$treatment <- factor(plot$treatment, levels = level_cell)

## plot average expression of genes across each cluster at each timepoint
ggplot(plot, aes(x=treatment, y=expression,fill=treatment))+
  geom_bar(stat = "identity", position = "dodge2") +
  facet_grid(.~gene) +
  scale_fill_manual(values = my.color.neural[c(1, 3, 5, 7, 9, 11)]) +
  theme_bw(base_size = 20) +
  theme(legend.position = "none") +
  Seurat:::RotatedAxis() +
  xlab("Cell Type") +
  ylab("Average Expression")

```

### violin plots
```{r vln pbx3 meis}
DefaultAssay(pax.all) <- "RNA"
### vln plot of meis1 and pbx3 from all
all.expression <- FetchData(pax.all, c("ident", "AK_Names","CellClass", "condition", "pbx3", "meis1"))

pp <- ggplot(all.expression, aes(x=CellClass, y=pbx3, fill=CellClass)) +
  geom_violin() +
  geom_jitter(size=0.3) +
  theme_bw(base_size = 30) +
  theme(legend.position = "none", axis.text = element_text(colour = "black")) +
    scale_fill_manual(values = my.color.neural[c(1, 3, 5, 7, 9, 11)]) +
  xlab("")

pm <- ggplot(all.expression, aes(x=CellClass, y=meis1, fill=CellClass)) +
  geom_violin() +
  geom_jitter(size=0.3) +
  theme_bw(base_size = 30) +
  scale_fill_manual(values = my.color.neural[c(1, 3, 5, 7, 9, 11)]) +
  theme(legend.position = "none", axis.text = element_text(colour = "black")) +
    xlab("")

plot_grid(pm, pp, nrow = 2)

```




## Supp 4E: Feature plot of pbx3 and meis1 in all cells
```{r supp 4d, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE}

# set assay
DefaultAssay(pax.all) <- "RNA"
Idents(pax.all) <- "AK_Names"

p1 <- FeaturePlot(pax.all, features = c("pbx3"),  order = TRUE, cols = c("dodgerblue3", "yellow", "forestgreen"))+ NoAxes()+NoLegend()
p2 <- FeaturePlot(pax.all, features = c("meis1"),  order = TRUE, cols = c("dodgerblue3", "yellow", "forestgreen"))+ NoAxes()+NoLegend()

plot_grid(p2, p1, nrow = 2)
```

## Supp 3F: Aztekin et al neural cell distinction
```{r aztekin, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE}

Idents(pax.neural) <- pax.neural$Laevis

neuralgenes <- c("mbp", "lepr", "lep", "foxa1", "foxa2", "tph1", "th", "gata3", "gata2", "tubb3", "mmp17", "mmp15", "tlx3", "tlx1", "pou4f1", "vsx2", "chat", "isl1", "mnx1", "lhx1", "pax2", "evx2", "evx1", "neurod4", "neurod1", "ascl1", "olig2", "dbx1", "irx3", "nkx6-1", "pax6", "nkx6-2", "arx")

DefaultAssay(pax.neural) <- "RNA"
DotPlot(pax.neural, features=neuralgenes, cols = c("blue", "gold")) + RotatedAxis() 

```

## Supp Figure 4: glia markers vs pan-nsc cells markers
```{r glia markers, echo=FALSE}

# Write function to blend feature plot with sox2
sox2Blend <- function(gene){
  FeaturePlot(pax.neural, blend = TRUE, features = c("sox2", gene), cols = c("#CC6600", "#009999"), order = TRUE, pt.size = 0.5)
}
```

### sox2 x elavl4
```{r elavl4, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE, fig.height=2, fig.width=8}
p1 <- sox2Blend("elavl4") # neg control neuron
p1
```

### sox2 x pax6
```{r pax6,cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE, fig.height=2, fig.width=8}
p2 <- sox2Blend("pax6") # positive control all NSC
p2
```

### sox2 x foxa1
```{r foxa1,cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE, fig.height=2, fig.width=8}
p3 <- sox2Blend("foxa1") # floor plate 1
p3
```

### sox2 x col8a1
```{r col8a1, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE, fig.height=2, fig.width=8}
p4 <- sox2Blend("col8a1") # floor plate 2
p4
```

### sox2 x hes1
```{r hes1, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE, fig.height=2, fig.width=8}
p5 <- sox2Blend("hes1") # radial glia marker 1
p5
```

### sox2 x vim
```{r vim,cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE, fig.height=2, fig.width=8}
p6<- sox2Blend("vim") # radial glia marker 2
p6
```

### sox2 x fabp7
```{r fabp7, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE, fig.height=2, fig.width=8}
p7 <- sox2Blend("fabp7") # radial glia marker 3 (blbp)
p7
```

### sox2 x slc1a3
```{r slc1a3, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE, fig.height=2, fig.width=8}
p8 <- sox2Blend("slc1a3") # radial glia marker 4 (glast)
p8
```

### plot
```{r plot glia markers,cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE}

plot_grid(p1, p2, p4, p3, ncol = 1)
plot_grid(p5, p6, p7, p8, ncol = 1)


```


