---
title: "Single Cell Figures"
author: "Anneke Kakebeen"
date: "13 Aug 2019"
output:
  html_document:
    code_folding: "hide"
    toc: true
    toc_depth: 3
    toc_float: true
fig_width: 6
fig_height: 4
editor_options: 
  chunk_output_type: console
---

## Set up for scRNA-Seq Figures
```{r scrnaseq figure set up, cache=FALSE, message=FALSE, warning=FALSE}
# load packages
library(Seurat)
library(LaCroixColoR)
library(ggplot2)
library(cowplot)
library(tidyr)
library(DT)
library(dplyr)
library(ggpubr)
library(rmarkdown)
library(knitr)

# load functions
source("~/Desktop/pax6 paper/Final Markdowns/common_source_functions.R")

# Set colors
my.color.neural <- c("#2CB11B", "#E9E4A6", "#DFCEE0" , "#E9A17C", "#1BB6AF" , "#172869" , "#FF3200")
condition.color <- c( "#6F0909", "#C0C0C0")
cc_colors <- c((lacroix_palette("Lime"))[c(1,3)], "black")

# Set working directory
setwd("~/Desktop/pax6 paper/Final Markdowns/SingleCell/analysis/")

```



## Read in scRNASeq Objects
```{r load objects, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE}

# Read in pax.neural seurat object
pax.neural <- readRDS("~/Desktop/pax6 paper/Final Markdowns/SingleCell/UMAP/pax.neural_FINAL.RDS")

# read in plotting data
pax.neural.plotting<- read.table("~/Desktop/pax6 paper/Final Markdowns/SingleCell/UMAP/pax.neural_plotting.txt", sep = "\t")

# factor plotting table for plotting
pax.neural.plotting$AK_Names <- factor(pax.neural.plotting$AK_Names, levels = c("NSC", "Transition", "Neural.A","Neural.B", "Neural.C", "Neural.D", "Prdm14" ))
pax.neural.plotting$condition <- factor(pax.neural.plotting$condition, levels = c("uninj", "dmso", "iwr"))
pax.neural.plotting$three_state <- factor(pax.neural.plotting$three_state, levels = c("NSC", "Differentiating", "Neuron"))
pax.neural.plotting$Laevis <- factor(pax.neural.plotting$Laevis, levels = c("Spinal Cord Progenitor", "Differentiating Neuron" , "Interneurons",  "Motor Neuron" , "Vulnerable Motor Neuron", "Motor Neuron (leptin+)" , "Dopaminergic Neurons" ))
pax.neural.plotting$Phase <- factor(pax.neural.plotting$Phase, levels = c("S", "G2M", "G1"))
```


## Figure 4a: UMAP of all neural cells
```{r 4a, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE}

# change idents of suerat object
Idents(pax.neural) <- pax.neural$Laevis
# plot wit hseurat umap
DimPlot(pax.neural, cols = my.color.neural) + NoAxes() + NoLegend()


```



## Figure 4b: UMAP of all neural cells colored for condition (uninjured or 24hpa)
```{r 4b, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE}
# change idents of suerat object
condition.neural <- subset(pax.neural, condition %in% c("uninj", "dmso"))
Idents(condition.neural) <- condition.neural$condition
DimPlot(condition.neural, cols = condition.color) + NoAxes() + NoLegend()

# ggplot(pax.neural.plotting, aes(x=UMAP_1, y=UMAP_2, color=pax.neural.plotting$mock_24))+
#   geom_point(stat="identity", aes(alpha = 0.6)) + 
#   scale_color_manual(values = condition.color) + 
#   Seurat:::NoAxes() + 
#   blank_theme +
#   theme(legend.position = "none")

```


## Figure 4C: Marker dotplot for neural cell clusters
```{r 4C, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE, fig.height= 6, fig.width=10}

# read in marker table
DEClusters.neural <- read.table("~/Desktop/pax6 paper/Final Markdowns/SingleCell/analysis/outs/DEClusters.neural.txt", sep = "\t", header =TRUE, row.names = 1)
# reassign class of genes to factors
DEClusters.neural$gene <- as.character(DEClusters.neural$gene)
# inspect table
datatable(DEClusters.neural)
# get
markers <- unique(c("sox2", (filter(DEClusters.neural, cluster =="DEClusters.NSC")[1:3,7]),
             (filter(DEClusters.neural, cluster == "DEClusters.Transition")[1:3,7]),
             (filter(DEClusters.neural, cluster == "DEClusters.Neural.A")[1:3,7]),
             (filter(DEClusters.neural, cluster == "DEClusters.Neural.B")[1:3,7]),
             (filter(DEClusters.neural, cluster == "DEClusters.Neural.C")[1:3,7]),
             (filter(DEClusters.neural, cluster == "DEClusters.Neural.D")[1:3,7]),
             "lepr",
              (filter(DEClusters.neural, cluster == "DEClusters.Prdm14")[1:3,7]),
             "prdm14", "prdm1"
             ))


DefaultAssay(pax.neural) <- "RNA" # change to RNA assay
pax.neural$AK_Names <- factor(pax.neural$AK_Names, levels = c("NSC", "Transition", "Neural.A", "Neural.B", "Neural.C", "Neural.D", "Prdm14")) # factor names in AK_Names for ordering purposes
Idents(pax.neural) <- pax.neural$Laevis # change object identity to AK_Names
pax.neural$Laevis <- factor(pax.neural$Laevis, levels = c("Spinal Cord Progenitor", "Differentiating Neuron", "Interneurons", "Vulnerable Motor Neuron", "Motor Neuron", "Motor Neuron (leptin+)", "Dopaminergic Neurons"))
Idents(pax.neural) <- factor(Idents(pax.neural), levels = (levels(pax.neural$Laevis))) # Reverse the order of names fo the plot

# plot dotplot
DotPlot(pax.neural, features = (markers), cols = c("#172869", "#F7AA14"))  + RotatedAxis() +  NoLegend() + coord_flip()

```


## Figure 4D-G: cell cluster proportions 
### Set up data table
```{r figure 4 cell prop, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE}

Idents(object = pax.neural) <- "threestate_condition_cell"

# Get number of cells in each cluster for each condition into data frame
nCells.neural <- as.data.frame(table(pax.neural@active.ident))
# Make condition and cluster columns
nCells.neural$cluster <- gsub(nCells.neural$Var1, pattern = ".*__", replacement = "")
nCells.neural$condition <- gsub(nCells.neural$Var1, pattern = "__.*", replacement = "")
nCells.neural$condition <- gsub(nCells.neural$condition, pattern = "_.*", replacement = "")
nCells.neural$class <- gsub(nCells.neural$Var1, pattern = "__.*", replacement = "")
nCells.neural$class <- gsub(nCells.neural$class, pattern = ".*_", replacement = "")
nCells.neural <- nCells.neural[order(nCells.neural$condition),]

# get total numner of cells for each condition
totals <-as.data.frame(nCells.neural %>%
    group_by(condition) %>% 
  summarise(totals = sum(Freq)))
totals$totals <- as.numeric(totals$totals)

# Make ratio columns in new table expressing percent cluster of all neural cells
nCells.neural <- merge.data.frame(nCells.neural, totals, by = "condition")
nCells.neural$percent <- nCells.neural$Freq/nCells.neural$totals*100

# mutate for sunburst
nCells.neural <- nCells.neural %>% mutate(id = seq(class))

# factor table
nCells.neural$condition <- factor(nCells.neural$condition, levels = c("uninj", "dmso", "iwr"))
nCells.neural$cluster <- factor(nCells.neural$cluster, levels = c("Spinal Cord Progenitor", "Differentiating Neuron", "Interneurons", "Vulnerable Motor Neuron", "Dopaminergic Neurons", "Motor Neuron (leptin+)", "Motor Neuron"))
nCells.neural$class <- factor(nCells.neural$class, levels = c("NSC", "Differentiating", "Neuron"))

```

### Pie Charts
```{r 4f/g uninjured cell proportions, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE}

# color for sunburst
my.color.neural.sun <- c("grey", "#E9E4A6", "#1BB6AF","#DFCEE0" , "#FF3200", "#172869" , "black", "white","#2CB11B", "#E9A17C")

# Filter table for unjured
plot.table <- filter(nCells.neural, condition %in% c("uninj", "dmso"))

plot.table.1 <- filter(nCells.neural, condition == "uninj")
p1 <- ggplot(plot.table.1, aes(y = percent, group=class)) +
  geom_col(aes(fill = class, x = 0), width = .4, color="black") +
  geom_col(aes(fill = cluster, x = .25), width = .1, color="black") +
  coord_polar(theta = 'y') +
  blank_theme +
  scale_fill_manual(values = my.color.neural.sun) +
  theme(legend.position = "none", axis.text = element_text(colour = "white"))

plot.table.2 <- filter(nCells.neural, condition == "dmso")
p2 <- ggplot(plot.table.2, aes(y = percent, group = class)) +
  geom_col(aes(fill = class, x = 0), width = .4, color="black") +
  geom_col(aes(fill = cluster, x = .25), width = .1, color="black") +
  coord_polar(theta = 'y') +
  blank_theme +
  scale_fill_manual(values = my.color.neural.sun) +
  theme(legend.position = "none", axis.text = element_text(colour = "white"))

plot_grid(p1, p2, ncol = 2, labels = c("uninjured", "24hpa"))

plot.table %>% group_by(condition, cluster) %>% summarise(sum = sum(percent))

# sunburst code
# (https://stackoverflow.com/questions/55263155/follow-up-of-how-to-make-a-sunburst-plot-in-r)
  

```

### UMAPS
```{r 4d/e uninjured cell proportions 2, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE}

# umap plot of cell types faceted by condition
ggplot(filter(pax.neural.plotting, condition %in% c("uninj", "dmso")), aes(x=UMAP_1, y=UMAP_2, color=AK_Names))+
  geom_point(stat="identity") +
  scale_color_manual(values = my.color.neural) +
  facet_grid(.~condition) +
  blank_theme +
  theme(legend.position = "right", axis.text = element_text(colour = "white"))


```

## Figure 4E: Cell cycle phase prediction
### Table set up
```{r fig4cell cycle, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE}

Idents(pax.neural) <- "phase_condition_cell"

# Get number of cells in each cluster for each condition into data frame
nCells.neural <- as.data.frame(table(pax.neural@active.ident))
# Make condition and cluster columns
nCells.neural$cluster <- gsub(nCells.neural$Var1, pattern = ".*__", replacement = "")
nCells.neural$condition <- gsub(nCells.neural$Var1, pattern = "__.*", replacement = "")
nCells.neural$condition <- gsub(nCells.neural$condition, pattern = "_.*", replacement = "")
nCells.neural$phase <- gsub(nCells.neural$Var1, pattern = "__.*", replacement = "")
nCells.neural$phase <- gsub(nCells.neural$phase, pattern = ".*_", replacement = "")
nCells.neural <- nCells.neural[order(nCells.neural$condition),]

# get total numner of cells for each condition
totals <-as.data.frame(nCells.neural %>%
                         group_by(condition) %>% 
                         summarise(totals = sum(Freq)))
totals$totals <- as.numeric(totals$totals)

nCells.neural <- merge.data.frame(nCells.neural, totals, by = "condition")
nCells.neural$percent <- nCells.neural$Freq/nCells.neural$totals*100

# mutate for sunburts
nCells.neural <- nCells.neural %>% mutate(id = seq(phase))

# factor table
nCells.neural$condition <- factor(nCells.neural$condition, levels = c("uninj", "dmso", "iwr"))
nCells.neural$cluster <- factor(nCells.neural$cluster, levels = c("Spinal Cord Progenitor", "Differentiating Neuron", "Interneurons", "Vulnerable Motor Neuron", "Dopaminergic Neurons", "Motor Neuron (leptin+)", "Motor Neuron"))
nCells.neural$phase <- factor(nCells.neural$phase, levels = c("S", "G2M", "G1"))


```

### Pie Charts
```{r 4j/k, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE}

# color for sunburst
my.color.neural.sun <- c("#E9E4A6", "#1BB6AF","black" , "#BDDE9B","#DFCEE0", "#FF3200", "#172869","#2CB11B" , "#2CB11B", "#E9A17C" )

# Filter table for unjured
plot.table <- filter(nCells.neural, condition %in% c("uninj", "dmso"))

# plot table
plot.table.1 <- filter(nCells.neural, condition == "uninj")
p1 <- ggplot(plot.table.1,
       aes(y = percent, group = phase)) +
  geom_col(aes(fill = phase, x = 0), width = .4) +
  geom_col(aes(fill = cluster, x = .25), width = .1, color="black") +
  coord_polar(theta = 'y') +
  blank_theme +
  scale_fill_manual(values = my.color.neural.sun) +
  theme(legend.position = "none", axis.text = element_text(colour = "white"))

plot.table.2 <- filter(nCells.neural, condition == "dmso")
p2 <- ggplot(plot.table.2,
       aes(y = percent, group = phase)) +
  geom_col(aes(fill = phase, x = 0), width = .4) +
  geom_col(aes(fill = cluster, x = .25), width = .1, color="black") +
  coord_polar(theta = 'y') +
  blank_theme +
  scale_fill_manual(values = my.color.neural.sun) +
  theme(legend.position = "none", axis.text = element_text(colour = "white"))

plot_grid(p1, p2, ncol = 2, labels = c("uninjured", "24hpa"))

plot.table %>% group_by(condition, phase) %>% summarise(sum = sum(percent))

```

### UMAP
```{r 4h/i charts, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE}

# umap plot of cell types faceted by condition
ggplot(filter(pax.neural.plotting, condition %in% c("uninj", "dmso")), aes(x=UMAP_1, y=UMAP_2, color=Phase))+
  geom_point(stat="identity") +
  scale_color_manual(values = cc_colors) +
  facet_grid(.~condition) +
  blank_theme +
  theme(legend.position = "right", axis.text = element_text(colour = "white"))


```


## Figure 5E: average expression of meis1 and pbx3 over sc clusters
```{r grn,cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE}

# Predicted GRN
" 6hpa_Gene:meis1 -> 24hpa_TF:meis1 -> 72hpa_TF:runx|etv1|klf9"
" 6hpa_Gene:etv1 -> 24hpa_TF:etv1 -> 72hpa_TF:pbx3"

# Define ins and outs for extraction function
Idents(pax.neural) <- "laevis_condition"
average_clusters <- t(as.data.frame(AverageExpression(pax.neural, assays = "RNA", use.counts = TRUE, features = c("meis1", "pbx3"))))
outs <- c("meis1_out", "pbx3_out")
genes <- c("meis1", "pbx3")

# Extract average expression data for plotting
seuratTOplot(average_clusters, gene = genes, outs = outs) # run to plot function (found in sourced functions)
objects <- ls(pattern=".*_out") # get all object names made in previous function
objects.list <- lapply(objects, get) # get outs
plot <- do.call(bind_rows, objects.list) # rowbind both gene data sets
# Make metdata columns
plot$treatment <- gsub(plot$condition, pattern = "RNA.", replacement = "") # split condition column
plot <- separate(plot, col = treatment, sep = "_", into = c("treament", "celltype")) # make seperate columns for treatment and celltype
plot$treament <- gsub(plot$treament, pattern = "dmso", replacement = "24hpa") # rename dmso to 24hpa

# Factor data to order
plot$gene <- factor(plot$gene, levels = c("meis1", "pbx3")) # order gene row for plot
plot$treament <- factor(plot$treament, levels = c("uninj", "24hpa", "iwr")) # order conditions
plot$celltype <- factor(plot$celltype, levels = c("Spinal.Cord.Progenitor", "Differentiating.Neuron", "Interneurons", "Vulnerable.Motor.Neuron", "Dopaminergic.Neurons", "Motor.Neuron..leptin..", "Motor.Neuron"))  

## plot average expression of genes across each cluster at each timepoint
time.plot <- filter(plot, treament %in% c("uninj", "24hpa"))
ggplot(time.plot, aes(x=treament, y=expression, group=treament, fill=celltype))+
  geom_bar(stat = "identity", position = "dodge2") +
  scale_fill_manual(values = my.color.neural) +
  theme_bw() +
  facet_grid(gene~celltype) +
  theme(legend.position = "none", 
        text = element_text(size=20), 
        axis.text.x = element_text(size = 30), 
        axis.title.y = element_text(size = 25), 
        axis.text = element_text(colour = "black", size = 30), 
        strip.background = element_blank(), 
        strip.text = element_blank()) +
  Seurat:::RotatedAxis()

Idents(pax.neural) <- pax.neural$Laevis
VlnPlot(pax.neural, features = c("meis1"), split.by = "mock_24", log = TRUE)
VlnPlot(pax.neural, features = c("pbx3"), split.by = "mock_24", log = TRUE)
```

```{r meis1 pbx3 neural}

# get data from seurat for violin plot
vlndata <- FetchData(pax.neural, c("ident", "Laevis", "mock_24", "pbx3", "meis1"))


plot_grid(pm, pp, nrow = 2)

### 2 violin
pm <- ggplot(vlndata, aes(x=Laevis, y=meis1, fill=mock_24)) +
  geom_violin() +
  geom_jitter(size=0.2) +
  theme_bw(base_size = 20) +
  scale_fill_manual(values = c("#c51b8a","#B574A7")) +
  xlab("")+
  ylab("meis1 expression")+
  theme(legend.position = "none", axis.text = element_text(colour = "black"), axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(), strip.background = element_blank(), strip.text = element_blank())

pp <- ggplot(vlndata, aes(x=Laevis, y=pbx3, fill=mock_24)) +
  geom_violin() +
  geom_jitter(size=0.2) +
  theme_bw(base_size = 20) +
  scale_fill_manual(values =  c("#e6550d", "#FBB58F")) +
  xlab("")+
  ylab("pbx3 expression")+
  theme(legend.position = "none", axis.text = element_text(colour = "black"), axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(), strip.background = element_blank(), strip.text = element_blank())

### split violin
pp <- ggplot(vlndata, aes(x=Laevis, y=pbx3, fill=mock_24)) +
  geom_split_violin() +
  geom_jitter(size=0.2) +
  theme_bw(base_size = 20) +
  scale_fill_manual(values =  c("#e6550d", "#FBB58F")) +
  xlab("")+
  ylab("pbx3 expression")+
  theme(legend.position = "none", axis.text = element_text(colour = "black"), axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(), strip.background = element_blank(), strip.text = element_blank())

pm <- ggplot(vlndata, aes(x=Laevis, y=meis1, fill=mock_24)) +
  geom_split_violin() +
  geom_jitter(size=0.2) +
  theme_bw(base_size = 20) +
  scale_fill_manual(values = c("#c51b8a","#B574A7")) +
  xlab("")+
  ylab("meis1 expression")+
  theme(legend.position = "none", axis.text = element_text(colour = "black"), axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(), strip.background = element_blank(), strip.text = element_blank())

```





## Figure 5F: sunburst of number of cells that express pbx3 and meis 1
```{r 5f, cache=TRUE, warning=FALSE, message=FALSE, echo=FALSE}

# get pax meta data
pax.meta <- pax.neural[[]] # get datat
pax.meta$cell <- rownames(pax.meta) # add colnames of cell id

# get TRUE/FALSE if cell is expressing genes of interets
# pbx3
pbx3.neural <- as.data.frame(GetAssayData(object = pax.neural)["pbx3",] >0)
colnames(pbx3.neural) <- "pbx3"
pbx3.neural$cell <- rownames(pbx3.neural)
#meis1
meis1.neural <- as.data.frame(GetAssayData(object = pax.neural)["meis1",] >0)
colnames(meis1.neural) <- "meis1"
meis1.neural$cell <- rownames(meis1.neural)

# merge meta and gene data frames
pbx3.neural <- merge.data.frame(pax.meta, pbx3.neural, by = "cell") #pbx3
meis1.neural <- merge.data.frame(pax.meta, meis1.neural, by = "cell") #meis1

# get frequency of cells expressing or not expressing gene
#pbx3
pbx3.neural.plot <- as.data.frame(table(pbx3.neural$Laevis, pbx3.neural$pbx3, pbx3.neural$condition))
colnames(pbx3.neural.plot) <- c("cluster", "expression", "condition", "frequency")
#meis1
meis1.neural.plot <- as.data.frame(table(meis1.neural$Laevis, meis1.neural$meis1, meis1.neural$condition))
colnames(meis1.neural.plot) <- c("cluster", "expression", "condition", "frequency")

# get total number of cells per condition
total.pbx <- as.data.frame(table(pbx3.neural$condition)) #pbx3
total.meis <- as.data.frame(table(meis1.neural$condition))#meis1

# merge total and gene data frames and make percentage column
# pbx
pbx3.neural.plot <- merge.data.frame(pbx3.neural.plot, total.pbx, by.x = "condition", by.y = "Var1")
pbx3.neural.plot$percent <- pbx3.neural.plot$frequency/pbx3.neural.plot$Freq*100
pbx3.neural.plot <- pbx3.neural.plot %>% mutate(id = seq(expression))
pbx3.neural.plot <- filter(pbx3.neural.plot, condition %in% c("dmso", "uninj"))
#meis
meis1.neural.plot <- merge.data.frame(meis1.neural.plot, total.pbx, by.x = "condition", by.y = "Var1")
meis1.neural.plot$percent <- meis1.neural.plot$frequency/meis1.neural.plot$Freq*100
meis1.neural.plot <- meis1.neural.plot %>% mutate(id = seq(expression))
meis1.neural.plot <- filter(meis1.neural.plot, condition %in% c("dmso", "uninj"))

 
plotcolors <- c("#E9E4A6", "#1BB6AF", "black", "#DFCEE0","#FF3200",  "#172869" ,"#2CB11B", "white",  "#E9A17C" )

# plot pbx3
ggplot(pbx3.neural.plot, 
       aes(y = percent, group = expression)) +
  geom_col(aes(fill = expression, x = 0), width = .4) + 
  geom_col(aes(fill = cluster, x = .25), width = .1, color="black") +
  coord_polar(theta = 'y') + 
  blank_theme +
  facet_grid(.~condition)+
  scale_fill_manual(values = plotcolors) +
  theme(legend.position = "none", axis.text = element_text(colour = "white")) +
  ggtitle("pbx3")

# plot meis1
ggplot(meis1.neural.plot, 
       aes(y = percent, group = expression)) +
  geom_col(aes(fill = expression, x = 0), width = .4) + 
  geom_col(aes(fill = cluster, x = .25), width = .1, color="black") +
  coord_polar(theta = 'y') + 
  blank_theme +
  facet_grid(.~condition)+
    scale_fill_manual(values = plotcolors) +
  theme(legend.position = "none", axis.text = element_text(colour = "white")) +
  ggtitle("meis1")

meis1.neural.plot %>% group_by(condition, expression) %>% summarise(sum = sum(percent))
pbx3.neural.plot %>% group_by(condition, expression) %>% summarise(sum = sum(percent))


```